### PR TITLE
Add one best CTC decoding

### PR DESCRIPTION
**Pull request by csukuangfj**
The motivation is to use k2 for CTC decoding for models trained using CTC loss from various frameworks, such as
- [icefall](https://github.com/k2-fsa/icefall)
- [torchaudio](https://github.com/pytorch/audio)
- [espnet](https://github.com/espnet/espnet)
- [speechbrain](https://github.com/speechbrain/speechbrain)
- [NeMo](https://github.com/NVIDIA/NeMo)
- [fairseq](https://github.com/facebookresearch/fairseq)
- [wenet](https://github.com/wenet-e2e/wenet)
- etc

Advantages to use k2:
 - Support batch
 - Support GPU
 - Support HLG graph
 - n-gram LM rescoring

This PR implements 1-best CTC decoding.

---

Code in this PR is mostly copied from
https://github.com/k2-fsa/icefall/blob/master/icefall/decode.py

### CODE REVIEW

**Comment by csukuangfj** [comment]
The following shows an example about how to use a pretrained model from wenet with CTC decoding using this PR.


```
wget https://github.com/wenet-e2e/wenet/releases/download/v2.0.1/en.tar.gz
tar xvf en.tar.gz
ls -lh en

$ ls -lh en
total 223M
-rwxr-xr-x 1 kuangfangjun root 222M Apr  3  2022 final.zip
-rwxr-xr-x 1 kuangfangjun root 1.7K Apr  3  2022 train.yaml
-rwxr-xr-x 1 kuangfangjun root  70K Apr  3  2022 units.txt
```

You only need `final.zip` and `units.txt` with this PR.
If you want to use HLG (i.e., TLG), you also need `HLG.pt`.

The following is the code for CTC decoding using the above-pretrained model:

[test.py.txt](https://github.com/k2-fsa/k2/files/9967334/test.py.txt)

```python3
#!/usr/bin/env python3

import math
from typing import List

import k2
import kaldifeat
import torch
import torchaudio
from torch.nn.utils.rnn import pad_sequence


def construct_fbank_computer(device: torch.device):
    print("Constructing Fbank computer")
    opts = kaldifeat.FbankOptions()
    opts.device = device
    opts.frame_opts.dither = 0
    opts.frame_opts.snip_edges = False
    opts.frame_opts.samp_freq = 16000
    opts.mel_opts.num_bins = 80

    fbank = kaldifeat.Fbank(opts)
    return fbank


def read_sound_files(filenames: List[str]) -> List[torch.Tensor]:
    ans = []
    for f in filenames:
        wave, sample_rate = torchaudio.load(f)
        assert sample_rate == 16000, (sample_rate, 16000)
        # We use only the first channel
        ans.append(wave[0].contiguous() * 32768)
    return ans


@torch.no_grad()
def main():
    device = torch.device("cpu")
    if torch.cuda.is_available():
        device = torch.device("cuda:0")

    m = torch.jit.load("final.zip").to(device)
    m.eval()
    fbank = construct_fbank_computer(device)

    # You can download them from
    # https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/tree/main/test_wavs
    # or use your own.
    filenames = ["./1089-134686-0001.wav", "./1221-135766-0001.wav"]

    waves = read_sound_files(filenames)
    waves = [w.to(device) for w in waves]

    features = fbank(waves)
    feature_lengths = [f.size(0) for f in features]

    features = pad_sequence(
        features,
        batch_first=True,
        padding_value=math.log(1e-10),
    )
    feature_lengths = torch.tensor(feature_lengths, device=device)

    encoder_out, mask = m.encoder(features, feature_lengths)
    log_prob = m.ctc.log_softmax(encoder_out)
    log_prob_len = mask.sum(dim=(1, 2))

    H = k2.ctc_topo(
        max_token=m.vocab_size - 1,
        modified=True,
        device=device,
    )

    lattice = k2.get_lattice(
        log_prob,
        log_prob_len,
        H,
        subsampling_factor=m.subsampling_rate(),
    )

    best_path = k2.one_best_decoding(lattice=lattice, use_double_scores=True)
    token_ids = k2.get_aux_labels(best_path)
    symbol_table = k2.SymbolTable.from_file("./units.txt")
    texts = [[symbol_table[i] for i in ids] for ids in token_ids]
    texts = ["".join(t) for t in texts]
    texts = [t.replace("▁", " ").strip() for t in texts]
    for filename, text in zip(filenames, texts):
        print(f"filename: {filename}\nresult: {text}\n\n")


if __name__ == "__main__":
    main()
```

**Comment by csukuangfj** [comment]
[test-torchaudio.py.txt](https://github.com/k2-fsa/k2/files/9968146/test-torchaudio.py.txt)

Here is another example using pre-trained models from [torchaudio](https://github.com/pytorch/audio)

```python3
#!/usr/bin/env python3

# see
# https://pytorch.org/audio/main/tutorials/asr_inference_with_ctc_decoder_tutorial.html

import torch
import k2
from typing import List
from torch.nn.utils.rnn import pad_sequence

import torchaudio


def read_sound_files(
    filenames: List[str],
    expected_sample_rate,
) -> List[torch.Tensor]:
    ans = []
    for f in filenames:
        wave, sample_rate = torchaudio.load(f)
        assert sample_rate == expected_sample_rate, (sample_rate, expected_sample_rate)
        # We use only the first channel
        ans.append(wave[0].contiguous())
    return ans


@torch.no_grad()
def main():
    bundle = torchaudio.pipelines.WAV2VEC2_ASR_BASE_10M
    acoustic_model = bundle.get_model()

    # You can download them from
    # https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/tree/main/test_wavs
    # or use your own.
    filenames = ["./1089-134686-0001.wav", "./1221-135766-0001.wav"]

    waves = read_sound_files(filenames, bundle.sample_rate)
    waves_len = [w.size(0) for w in waves]
    waves_len = torch.tensor(waves_len)
    padded_waves = pad_sequence(
        waves,
        batch_first=True,
        padding_value=0,
    )

    log_prob, log_prob_len = acoustic_model(padded_waves, waves_len)
    log_prob = log_prob.log_softmax(-1)

    vocab_size = log_prob.size(-1)

    H = k2.ctc_topo(
        max_token=vocab_size - 1,
        modified=True,
        device="cpu",
    )

    lattice = k2.get_lattice(
        log_prob,
        log_prob_len,
        H,
        subsampling_factor=1,
    )

    best_path = k2.one_best_decoding(lattice=lattice, use_double_scores=True)
    token_ids = k2.get_aux_labels(best_path)

    symbols = bundle.get_labels()
    symbol_ids = list(range(len(symbols)))
    symbol_table = dict(zip(symbol_ids, symbols))

    texts = [[symbol_table[i] for i in ids] for ids in token_ids]
    texts = ["".join(t) for t in texts]
    texts = [t.replace("|", " ").strip() for t in texts]
    for filename, text in zip(filenames, texts):
        print(f"filename: {filename}\nresult: {text}\n\n")


if __name__ == "__main__":
    main()
```
